Lib/unittest/suite.py

```diff
@@ -150,34 +150,38 @@ def _handleClassSetUp(self, test, result):
             return
 
         try:
-            currentClass._classSetupFailed = False
+            failed = currentClass._classSetupFailed = False
         except TypeError:
             # test may actually be a function
             # so its class will be a builtin-type
             pass
 
         setUpClass = getattr(currentClass, 'setUpClass', None)
+        doClassCleanups = getattr(currentClass, 'doClassCleanups', None)
         if setUpClass is not None:
             _call_if_exists(result, '_setupStdout')
             try:
-                setUpClass()
-            except Exception as e:
-                if isinstance(result, _DebugResult):
-                    raise
-                currentClass._classSetupFailed = True
-                className = util.strclass(currentClass)
-                self._createClassOrModuleLevelException(result, e,
-                                                        'setUpClass',
-                                                        className)
+                try:
+                    setUpClass()
+                except Exception as e:
+                    if isinstance(result, _DebugResult):
+                        raise
+                    try:
+                        failed = currentClass._classSetupFailed = True
+                    except TypeError:
+                        pass
+                    className = util.strclass(currentClass)
+                    self._createClassOrModuleLevelException(result, e,
+                                                            'setUpClass',
+                                                            className)
+                if failed and doClassCleanups is not None:
```

Comment:
`failed` will raise `UnboundLocalError` if `currentClass` is a function (builtin-type). We need to give it a default.

Reply:
It is set at line 153. I tested this case.

Reply:
No, you also need to set it in line 157.

Reply:
In `failed = currentClass._classSetupFailed = False` first `failed` is set, then `currentClass._classSetupFailed` is tried to set.

I rewritten the code in more clear form.